src/avx512bw/avx512bw-from-valid-utf8.inl.cpp

```diff
@@ -0,0 +1,339 @@
+    }
+
+    uint16_t words[32];
+
+    {
+        // 2. build surrogate pair words in 32-bit lanes
+
+        //    t0 = 8 x [000000000000aaaa|aaaaaabbbbbbbbbb]
+        const __m512i t0 = _mm512_sub_epi32(utf32, v_0001_0000);
+
+        //    t1 = 8 x [000000aaaaaaaaaa|bbbbbbbbbb000000]
+        const __m512i t1 = _mm512_slli_epi32(t0, 6);
+
+        //    t2 = 8 x [000000aaaaaaaaaa|aaaaaabbbbbbbbbb] -- copy hi word from t1 to t0
+        //         0xe4 = (t1 and v_ffff_0000) or (t0 and not v_ffff_0000)
+        const __m512i t2 = _mm512_ternarylogic_epi32(t1, t0, v_ffff_0000, 0xe4);
+
+        //    t2 = 8 x [110110aaaaaaaaaa|110111bbbbbbbbbb] -- copy hi word from t1 to t0
+        //         0xba = (t2 and not v_fc00_fc000) or v_d800_dc00
+        const __m512i t3 = _mm512_ternarylogic_epi32(t2, v_fc00_fc00, v_d800_dc00, 0xba);
+
+        _mm512_storeu_si512((__m512i*)words, t3);
+    }
+
+    // 3. store valid 16-bit values
+    _mm256_mask_storeu_epi16((__m256i*)output, valid, _mm512_cvtepi32_epi16(utf32));
+
+    int sp = __builtin_popcount(sp_mask);
+
+    // 4. copy surrogate pairs
+    uint32_t mask = sp_mask;
+    for (int i=count; i >= 0 && mask != 0; i--) {
+        if (test_and_clear_bit(mask, i)) {
+            output[i + sp] = words[2*i + 0];
+            sp -= 1;
+            output[i + sp] = words[2*i + 1];
+        } else {
+            output[i + sp] = output[i];
+        }
+    }
+
+    return count + __builtin_popcount(sp_mask);
+}
+
+/*
+    valid_utf8_to_fixed_length converts a valid UTF-8 string into UTF-32.
+
+    The `OUTPUT` template type decides what to do with UTF-32: store
+    it directly or convert into UTF-16 (with AVX512).
+
+    Input:
+    - str           - valid UTF-8 string
+    - len           - string length
+    - out_buffer    - output buffer
+
+    Result:
+    - pair.first    - the first unprocessed input byte
+    - pair.second   - the first unprocessed output word
+*/
+
+template <typename OUTPUT>
+std::pair<const char*, OUTPUT*> valid_utf8_to_fixed_length(const char* str, size_t len, OUTPUT* out_buffer) {
+    constexpr bool UTF32 = std::is_same<OUTPUT, uint32_t>::value;
+    constexpr bool UTF16 = std::is_same<OUTPUT, char16_t>::value;
+    static_assert(UTF32 or UTF16, "output type has to be uint32_t (for UTF-32) or char16_t (for UTF-16)");
+
+    const char* ptr = str;
+    const char* end = ptr + len;
+
+    OUTPUT* output = out_buffer;
+
+    while (ptr + 16 + 4 < end) {
+        // 1. Load all possible 4-byte substring into an AVX512 register.
+        const __m512i input = expand_bytes(ptr);
+
+        /* 2. Classify which words contain valid UTF-8 characters.
+              We test if the 0th byte is not a continuation byte (0b10xxxxxx) */
+        __mmask16 valid;
+        {
+            const __m512i t0 = _mm512_and_si512(input, v_0000_00c0);
+            valid = _mm512_cmpneq_epu32_mask(t0, v_0000_0080);
+        }
+        const int valid_count = __builtin_popcount(valid);
```

Comment:
Note here that you are doing an implicit cast of valid. It could be better to be explicit and to use `_cvtmask16_u32`. That is, it is important to express the fact that each time you access the mask as a regular integer, there is a rather slow (latency-wise) process. The mask registers are not mapped for free. There is a 3 cycles latency!!!!!!

Reply:
I agree it's important. Thanks, added.